tools/tool-testing/run.js

```diff
@@ -0,0 +1,443 @@
+// Represents a test run of the tool (except we also use it in
+// tests/old.js to run Node scripts). Typically created through the
+// run() method on Sandbox, but can also be created directly, say if
+// you want to do something other than invoke the 'meteor' command in
+// a nice sandbox.
+//
+// Options: args, cwd, env
+//
+// The 'execPath' argument and the 'cwd' option are assumed to be standard
+// paths.
+//
+// Arguments in the 'args' option are not assumed to be standard paths, so
+// calling any of the 'files.*' methods on them is not safe.
+import { spawn } from 'child_process';
+import * as files from '../fs/files.js';
+import { markTop as parseStackMarkTop } from '../utils/parse-stack.js';
+import Matcher from './matcher.js';
+import OutputLog from './output-log.js';
+import { randomPort, timeoutScaleFactor, sleepMs } from '../utils/utils.js';
+import TestFailure from './test-failure.js';
+import { execFileSync } from '../utils/processes.js';
+
+let runningTest = null;
+
+export function setRunningTest(test) {
+  runningTest = test;
+}
```

Comment:
I'll call this out ahead of time as I expect this `setRunningTest` method (used by [`Test`'s `onCleanup`](https://github.com/meteor/meteor/blob/abernix/decompose-self-test/tools/tool-testing/selftest.js#L83-L85); registered in the `constructor` below) to be the most subject to scrutiny portion of this re-factor.  It had previously been a global variable within the lone `selftest.js` file but it could stand to be further refined in this re-worked edition of `self-test`.  Ultimately, I just didn't want to completely abandon the work I had done on this PR, even though the work which prompted it had to be put on hold.